rollup/sync_service.go

```diff
@@ -0,0 +1,844 @@
+		return s.ProcessQueueBatchAppendedLog(ctx, ethlog)
+	}
+
+	return fmt.Errorf("Unknown log topic %s", hexutil.Encode(topic))
+}
+
+func (s *SyncService) ProcessTransactionEnqueuedLog(ctx context.Context, ethlog types.Log) error {
+	event, err := s.ctcFilterer.ParseTransactionEnqueued(ethlog)
+	if err != nil {
+		// add context to error
+		return err
+	}
+
+	// Nonce is set by god key at execution time
+	// Value and gasPrice are set to 0
+	// nil is the txid (unused)
+	tx := types.NewTransaction(uint64(0), event.Target, big.NewInt(0), event.GasLimit.Uint64(), big.NewInt(0), event.Data, &event.L1TxOrigin, nil, types.QueueOriginL1ToL2, types.SighashEIP155)
+
+	// Timestamp is used to update the blockchains clocktime
+	timestamp := time.Unix(event.Timestamp.Int64(), 0)
+	rtx := RollupTransaction{tx: tx, timestamp: timestamp, blockHeight: ethlog.BlockNumber}
+	s.l1ToL2Enqueued.Store(event.QueueIndex.Uint64(), rtx)
+
+	return nil
+}
+
+func (s *SyncService) ProcessSequencerBatchAppendedLog(ctx context.Context, ethlog types.Log) error {
+	event, err := s.ctcFilterer.ParseSequencerBatchAppended(ethlog)
+	if err != nil {
+		return err
+	}
+
+	tx, pending, err := s.ethclient.TransactionByHash(ctx, ethlog.TxHash)
+	if err == ethereum.NotFound {
+		log.Error("Transaction not found", "hash", ethlog.TxHash.Hex())
+		return err
+	}
+	if err != nil {
+		log.Error("Cannot fetch transaction", "hash", ethlog.TxHash.Hex())
+		return err
+	}
+	if pending {
+		log.Error("Transaction unexpectedly in mempool", "hash", ethlog.TxHash.Hex())
+		return err
+	}
+
+	cd := appendSequencerBatchCallData{}
+	err = cd.Decode(bytes.NewReader(tx.Data()))
+	if err != nil {
+		return err
+	}
+
+	// event.StartingQueueIndex
+	// event.NumQueueElements
+
+	// event.TotalElements
+	for i, element := range cd.ChainElements {
+		var tx *types.Transaction
+		index := event.TotalElements.Uint64() - (uint64(i) + event.NumQueueElements.Uint64())
+		// Sequencer transaction
+		if element.IsSequenced {
+			nonce := uint64(0)
+			to := common.Address{} // sequencerDecompressionAddress
+			gasLimit := uint64(0)  // max gas limit
+			l1TxOrigin := common.Address{}
+			tx = types.NewTransaction(nonce, to, big.NewInt(0), gasLimit, big.NewInt(0), element.TxData, &l1TxOrigin, nil, types.QueueOriginSequencer, types.SighashEIP155)
+			// Check if there is already a transaction at the index
+			if block := s.bc.GetBlockByNumber(index); block != nil {
+				// Check if the transaction is different
+				// this check will not work, need to check the items in the tx
+				// instead.
+				if included := block.Transaction(tx.Hash()); included == nil {
+					previous := block.Transactions()[0]
+					log.Info("Different transaction detected, reorganizing", "new", tx.Hash().Hex(), "previous", previous.Hash().Hex())
+					err := s.bc.SetHead(index - 1)
+					if err != nil {
+						log.Error("Cannot reorganize")
+						return err
+					}
+				}
+			}
+		} else {
+			// Queue transaction
```

Comment:
This should also have the reorg logic, instead of checking the tx hash, it should check equality of:

```
to + data + l1TxOrigin + gasLimit
```